README
NAME
    PHP - embedded PHP interpreter

DESCRIPTION
    The module makes it possible to execute PHP code, call PHP functions and
    methods, manipulate PHP arrays, and create PHP objects.

SYNOPSIS
            use PHP;

    General use

            # evaluate arbitrary PHP code; exception is thrown
            # and can be caught via standard eval{}/$@ block 
            PHP::eval(<<EVAL);
            function print_val(\$arr,\$val) {
                    echo \$arr[\$val];
            }

            # evaluate PHP expressions
            $answer = PHP::eval_return('7*(1+2+3);');
        
            class TestClass {
                    function TestClass ( $param ) {}
                    function method(\$val) { return \$val + 1; }
            };
            EVAL

            # catch output of PHP code
            PHP::options( stdout => sub {
                    print "PHP says: $_[0]\n";
            });
            PHP::eval('echo 42;');

    Arrays, high level

            # create a php array
            my $array = PHP::array;

            # access pseudo-hash content
            $array-> [1] = 42;
            $array-> {string} = 43;
        
            # pass arrays to function
            # Note - function name is not known by perl in advance, and
            # is called via AUTOLOAD
            PHP::print_val($array, 1);
            PHP::print_val($array, 'string');

    Arrays, low level

            # create a php array handle
            my $array = PHP::ArrayHandle-> new();
            # tie it either to an array or a hash
            my ( @array, %hash);
            $array-> tie(\%hash);
            $array-> tie(\@array);

            # access array content
            $array[1] = 42;
            $hash{2} = 43;

    Objects and properties

            my $TestClass = PHP::Object-> new('TestClass');
            print $TestClass-> method(42), "\n";
        
            $TestClass-> tie(\%hash);
            # set a property
            $hash{new_prop} = 'string';

API
    eval $CODE
        Feeds embedded PHP interpreter with $CODE, throws an exception on
        failure.

    call FUNCTION ...
        Calls PHP function with list of parameters. Returns exactly one
        value.

    include, include_once, require, require_once
        Shortcuts to the identical PHP constructs.

    array [ $REFERENCE ]
        Returns a handle to a newly created "PHP::Array" object, which can
        be accessed both as array and hash reference:

                $_ = PHP::array;
                $_->[42] = 'hello';
                $_->{world} = '!';

        If $REFERENCE is a "PHP::ArrayHandle" instance, then the newly
        created object is a pheudo-hash alias to the PHP array behind the
        $REFERENCE. If no $REFERENCE is given, a new PHP array is created.

    PHP::Object->new($class_name, @parameters)
        Instantiates a PHP object of PHP class $class_name and returns a
        handle to it. The methods of the class can be called directly via
        the handle:

                my $obj = PHP::Object-> new( 'MyClass', @params_to_constructor);
                $object-> method( @some_params);

        The relevant class constructor is called, if available, according to
        PHP specification, that is different between v4 and v5. The v4
        constructor has identical name with the class name; the v5
        constructor can also be named "__construct".

    PHP::Entity->tie($array_handle, $tie_to)
        Ties existing handle to a PHP entity to either a perl hash or a perl
        array. The tied hash or array can be used to access PHP pseudo_hash
        values indexed either by string or integer value.

        The PHP entity can be either an array, represented by
        "PHP::ArrayHandle", or an object, represented by "PHP::Object". In
        the latter case, the object properties are represented as hash/array
        values.

    PHP::Entity->link($original, $link)
        Records a reference to an arbitrary perl scalar $link as an alias to
        $original "PHP::Entity" object. This is used internally by
        "PHP::TieHash" and "PHP::TieArray", but might be also used for other
        purposes.

    PHP::Entity::unlink($link)
        Removes association between a "PHP::Entity" object and $link.

    PHP::Array->tie($self, $tie_to)
        Same as PHP::Entity->tie, but operates on "PHP::Array" objects.

    PHP::Array->handle
        Returns PHP array handle, a "PHP::ArrayHandle" object.

    PHP::options
        Contains set of internal options. If called without parameters,
        returns the names of the options. If called with a single parameter,
        return the associated value. If called with two parameters, replaces
        the associated value.

        debug $integer
            If set, loads of debugging information are dumped to stderr

            Default: 0

        stdout/stderr $callback
            "stdout" and "stderr" options define callbacks that are called
            when PHP decides to print something or complain, respectively.

            Default: undef

        version
            Read-only option; returns the version of PHP library compiled
            with .

DEBUGGING
    Environment variable "P5PHPDEBUG", if set to 1, turns the debug mode on.
    The same effect can be achieved programmatically by calling

            PHP::options( debug => 1);

INSTALLATION
    The module uses php-embed SAPI extension to inter-operate with PHP
    interpreter. That means php must be configured with '--enable-embed'
    parameters prior to using the module. Also, no '--with-apxs' must be
    present in to configuration agruments either, otherwise the PHP library
    will be linked with Apache functions, and will be unusable from the
    command line.

    The "sub dl_load_flags { 0x01 }" code in PHP.pm is required for PHP to
    load correctly its extensions. If your platform does RTLD_GLOBAL by
    default and croaks upon this line, it is safe to remove the line.

WHY?
    While I do agree that in general it is absolutely pointless to use PHP
    functionality from within Perl, scenarios where one must connect an
    existing PHP codebase to something else, are not something unusual.
    Also, this module might be handy for people who know PHP but are afraid
    of switching to Perl, or want to reuse their old PHP code.

    Currently, not all of PHP functionality is implemented, but OTOH I don't
    really expect this module to grow that big, because I believe it is
    easier to call "PHP::eval" rather than implement all the subtleties of
    Zend API. There are no callbacks to Perl from PHP code, and I don't
    think these are needed, because one thing is to be lazy and not to
    rewrite PHP code, and another is to make new code in PHP that uses Perl
    when PHP is not enough. As I see it, the latter would kill all incentive
    to switch to Perl, so I'd rather leave callbacks unimplemented.

SEE ALSO
    Using Perl code from PHP:
    <http://www.zend.com/php5/articles/php5-perl.php>

COPYRIGHT
    This library is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.

AUTHOR
    Dmitry Karasik <dmitry@karasik.eu.org>

 

Changes
Revision history for Perl extension PHP.

0.16  ...
        - RT#82890 Makefile.PL suppress cpantester report if PHP not available
	- RT#83345 Makefile.PL change for some debian systems

0.15  20 Jan 2013 (mob)
        - pass $replace arg to header callback

0.14  23 Jul 2012 (mob)
        - mechanism to update internal hash and make  is_uploaded_file()
          return true for files uploaded outside of PHP
	- mechanism to pass post content and set up php://input stream
        - fix PHP::assign_global for all superglobals

0.13  04 Mar 2012 (mob)
        - added PHP::__reset and PHP::assign_global functions
        - added PHP::options(header => ...) callback

0.12  10 Dec 2010
        - support PHP v5.2 and v5.3

0.11  09 Feb 2007
	- back-compat for php4

0.10  08 Dec 2006
	- reset EG(exception) so further php calls aren't aborted

0.09  24 May 2005
        - add PUSH/POP for array
        - fix TieArray::FETCHSIZE

0.08  02 May 2005
	- support PHPv5 constructors 
	- fix a memory leak

0.07  20 April 2005
	- re-implement reference counting scheme to fix some memory corruptions 

0.06  16 March 2005
        - return pseudo-hash from array-returning functions
	
0.05  2 March 2005
        - implement a pseudo-hash

0.04  24 Feb 2005
	- catch more exceptions

0.03  16 Feb 2005
	- compile with gcc versions below 3.3

0.02  15 Feb 2005
	- implement object properties and parameters to class constructors

0.01  12 Feb 2005
	- original version


test.pl
#$Id: test.pl,v 1.14 2007/02/11 10:59:14 dk Exp $

use Test::More tests => 76;
use strict;

# pre-test. bail out if PHP libraries are not installed on this system.



BEGIN {
    ok( eval "use PHP;1", 'use_ok PHP' );
    if ($@ && $@ =~ /Can't locate loadable object for module PHP/) {
	diag $@;
	BAIL_OUT(
	    "PHP libraries compiled for embedded SAPI not found" );
    }
}
require_ok('PHP');

# 3 
eval {
PHP::eval(<<'S1');

function loopback($var)
{
	return $var;
}

function new_array()
{
	return array();
}

function print_val($arr,$val)
{
	echo $arr[$val];
}

class TestClass
{
	var $prop;
	function TestClass ($a) { echo $a; }
	function method($val) { return $val + 1; }
	function getprop() { return $this->prop; }
};

S1
};
ok( !$@, 'eval');
die $@ if $@;

# 4,5
my $val;
eval {
	$val = PHP::call( 'loopback', 42);
};
ok( !$@, 'call');
ok( defined $val && $val eq '42', 'pass arguments, return values');

# 6
eval {
	PHP::eval('$');
};
ok( $@, 'invalid syntax exceptions');

# 7 
my $output = '';
PHP::options( stdout => sub { $output = shift});
PHP::eval( 'echo 42;');
ok( $output eq '42', 'catch output');

# 8 
my $a = PHP::new_array();
ok( $a, 'get array from php');

# 9
my $b = PHP::ArrayHandle-> new();
ok( $b, 'create array handle');

my ( @array, %hash);
$a->tie(\%hash);
$a->tie(\@array);

# 10
$array[1] = 'array';
ok( defined $array[1] && $array[1] eq 'array', 'tied array');

# 11
$hash{'h'} = 'hash';
ok( defined $hash{'h'} && $hash{'h'} eq 'hash', 'tied hash');

# 12
PHP::print_val($a, 1);
ok( $output eq 'array', 'query array value');

# 13
PHP::print_val($a, 'h');
ok( $output eq 'hash', 'query hash value');

# 14
PHP::TieHash::STORE( $b, '42', '42');
ok( PHP::TieHash::FETCH( $b, '42') eq '42', 'direct array access');

# 15
$output = '';
my $TestClass = PHP::Object-> new('TestClass', '43');
ok( $TestClass && $output eq '43', 'class');

# 16
ok( $TestClass-> method(42) == 43, 'methods');

# 17
$TestClass->tie(\%hash);
$hash{prop} = 42;
ok( $TestClass-> getprop() == 42, 'properties');

# 18
eval {
PHP::eval('call_unexistent_function_wekljfhv2kwfwkfvbwkfbvwjkfefv();');
};
ok($@ && $@ =~ /call_unexistent_function/, 'undefined function exceptions');

# 19
my $arr = PHP::array;
$arr->[1] = 42;
ok( $arr->[1] == 42, 'pseudo-hash, as array');

# 20
$arr->{'mm'} = 42;
ok( $arr->{'mm'} == 42, 'pseudo-hash, as hash');

# 21
my @k = keys %$arr;
ok(( 2 == @k and 2 == scalar grep { m/^(1|mm)$/ } @k), 'hash keys');

$output = '';
SKIP:{
	skip "php5 required", 3 unless PHP::options('version') =~ /^(\d+)/ and $1 > 4;
	eval { PHP::eval(<<'MOO'); };
class P5 {
	function __construct() { echo "CREATE"; }
	function __destruct() { echo "DESTROY"; }
}
function p5(){$a = new P5;return $a;}
MOO
	{
	my $P5 = PHP::Object-> new('P5');
	ok(!$@ && $P5, 'php5 syntax');
	ok($output eq 'CREATE', 'php5 constructors');
	}
	ok($output eq 'DESTROY', 'php5 destructors');
}

# 25
ok( scalar( @$arr) == 2, 'sparse arrays');

# 26
ok( 5 == push (@$arr, qw(1 2 3)), 'push');

# 27
my $k = 0 || pop @$arr; 
ok(( 4 == @$arr and '3' eq $k), 'pop');

undef $arr;


# 28
eval { PHP::eval('throw new Exception("bork");'); };
my $exc = $@;
$val = PHP::call( 'loopback', 42);
ok(( $exc and $val == 42), 'exceptions in eval');

# 29
SKIP:{
skip "php5 required", 1 unless PHP::options('version') =~ /^(\d+)/ and $1 > 4;
PHP::eval('function boom() { throw new Exception("bork"); } '); 
eval { PHP::call( 'boom'); };
my $exc = $@;
$val = PHP::call( 'loopback', 42);
ok(( $exc and $val == 42), 'exceptions in calls');
}

# 30
my @test30 = ();
PHP::eval('$foo = 43;');
PHP::options( stdout => sub { push @test30, $_[0] } );
PHP::eval('echo "hello " . $foo;');

PHP::__reset();
PHP::options( stdout => sub { push @test30, $_[0] } );
PHP::eval('echo "world " . $foo;');
ok(@test30 == 2 && $test30[0] eq 'hello 43' && $test30[1] eq 'world ',
    "PHP::__reset clears variables in previous instance");

# 31
PHP::assign_global("test31", 75);
PHP::eval('function test31() { global $test31; return $test31; }');
my $a31 = PHP::call('test31');
ok($a31 == 75, "assign_global simple scalar");

# 32
PHP::assign_global("test32", [1, 19, "qwert"]);
PHP::eval('function test32() { global $test32; return $test32; }');
my $a32 = PHP::call('test32');
ok(ref($a32) eq 'PHP::Array' && $a32->[0]==1 && $a32->[1]==19 && $a32->[2]eq'qwert',
    "assign_global listref");

# 33
PHP::assign_global("test33", { foo => [ 19, 52 ], cats => "the other white meat" });
PHP::eval('function test33() { global $test33; return $test33; }');
my $a33 = PHP::call('test33');
ok(ref($a33) eq 'PHP::Array' && $a33->{foo}[1]==52 && $a33->{cats} =~ /white meat/,
    "assign_global complex data structure");

# 34
my $r = PHP::eval_return('40+2');
ok( $r == 42, 'simple eval/return');

# 35
$r = PHP::eval_return('$test33');
ok( $r && ref($r) && ref($r) eq 'PHP::Array' && $r->{cats} =~ /white meat/,
    'global var in eval/return');

# 36
my @h36 = ();
PHP::options( header => sub { push @h36, $_[0]; } );
PHP::eval('header("Subject: Payload");');
PHP::eval_return('header("Header-X: This is a header",false);');
ok( $h36[0] eq 'Subject: Payload' && $h36[1] =~ /This is a header/,
    "PHP::options(header => ...) callback" );

# 37
my $a37 = [ 42 ];
my $b37 = [ $a37 ];
$a37->[0] = $b37;
$SIG{ALRM} = sub { die "Timeout assigning global with circular ref\n" };
alarm 5;
my $z37 = eval {
    PHP::assign_global( 'foo', $a37 );
    1;
};
alarm 0;
ok( $z37, 'assign_global with circular ref didn\'t cause infinite loop' );

# 38
PHP::assign_global( 'g38', bless ['I','II','III','IV'],'Roman::Numeral::Array' );
my $g38 = PHP::eval_return('$g38;');
ok( $g38 && ref($g38) && ref($g38) eq 'PHP::Array'
    && $g38->[0] eq 'I' && $g38->[3] eq 'IV',
    'assign_global blessed list reference' );

# 39
PHP::assign_global( 'g39', bless { Brazil => 'Brasilia', Uruguay => 'Montevideo' }, 'South::American::Capitols' );
my $g39 = PHP::eval_return( '$g39;' );
ok( $g39 && ref($g39) && ref($g39) eq 'PHP::Array'
    && $g39->{Brazil} eq 'Brasilia' && $g39->{Uruguay} eq 'Montevideo',
    'assign_global blessed hash reference' );

# 40
PHP::assign_global( 'g40', { hash => { abc => 'def' }, list => [3,4,5], scalar => 4 } );
my $g40 = PHP::eval_return( '$g40;' );
ok( $g40 && ref($g40) && ref($g40) eq 'PHP::Array'
    && ref($g40->{hash}) eq 'PHP::Array'
    && ref($g40->{list}) eq 'PHP::Array'
    && ref($g40->{scalar}) eq '',
    'assign_global: complex data structure has nested PHP::Array' );

# 41
ok( PHP::eval_return('"Evaluate PHP statement from Perl? Awesome!";') =~ /awesome/i,
    'eval/return simple string' );

# 42
PHP::eval('$foo=5;');
ok( PHP::eval_return('$foo*$foo;') == 25,
    'eval/return simple expression with variables ');

# 43-45
ok( PHP::eval_return('TRUE;') == 1, 'eval/return TRUE maps to 1' );
ok( PHP::eval_return('FALSE;') eq '', 'eval/return FALSE maps to ""' );
ok( !defined(PHP::eval_return('NULL;')), 'eval/return NULL maps to undef' );

# 46-47
ok( PHP::eval_return('4 > 5 ? 19 : 42;') == 42, 'eval/return ternary 1' );
ok( PHP::eval_return('4 < 5 ? 19 : 42;') == 19, 'eval/return ternary 2' );

# 48-53: the don'ts of eval_return
eval { PHP::eval_return('return "foo";') };
ok( $@, 'don\'t use "return" inside PHP::eval_return' );

eval { PHP::eval_return('if (2<17) 8; else 9;') };
ok ($@, 'don\'t use "if", "if/else" inside PHP::eval_return' );

eval { PHP::eval_return('function foo() { return 42; }') };
ok ($@, 'don\'t declare functions inside PHP::eval_return' );

eval { PHP::eval_return('echo 16;') };
ok ($@, 'don\'t use echo inside PHP::eval_return' );

eval { PHP::eval_return('die("die message");') };
ok ($@, 'don\'t call die() inside PHP::eval_return' );

eval { PHP::eval_return('exit(4);') };
ok ($@, 'don\'t call exit() inside PHP::eval_return' );

# 54-55
my $t54 = eval { PHP::eval_return(''); };
ok( !$@ && !defined($t54), 'eval/return(empty) ok, undef' );

my $t55 = eval { no warnings 'uninitialized'; PHP::eval_return(undef); };
ok( !$@ && !defined($t55), 'eval/return(undef) ok, undef' );

# 56-57
my $t56 = eval { PHP::eval_return('abs(-6)-log10(1000);') ; };
ok( $t56==3, 'eval/return with math functions' );

my $t57 = eval { PHP::eval_return("date('Y',$^T);"); };
ok($t57 >= 2012 && $t57 < 2020, 'eval/return with datetime functions');

# 58-60
my $t58 = eval { PHP::eval_return("array(5,10,15,20);"); };
ok($t58->[0] == 5 && $t58->[3] == 20, 'eval/return simple array');

my $t59 = eval { PHP::eval_return('array("abc"=>45, "ghi"=>"tennis");') };
ok($t59->{abc} == 45 && $t59->{ghi} eq "tennis" && !defined($t59->{foo}),
    'eval/return simple associative array');

my $t60 = eval { PHP::eval_return( q^
array(11, 16, array( "Perl" => "good", "PHP" => "meh" ),
      array(5, 8, 13, 21, 34));^) };
ok($t60->[1] == 16 && $t60->[2]{PHP} !~ /good/ && $t60->[3][3] == 21,
    'eval/return more complex data structure');

# 61-63
my $fake_file = "/fj234oi/453rerf3v434v3.txt";
my $fake_file2 = "/fj234o1e2123ei/453rqwrqe1241423rwerf3v434v3.txt";
ok( ! PHP::eval_return( "is_uploaded_file('$fake_file')" ),
    "is_uploaded_file false for fake file" );
PHP::_spoof_rfc1867($fake_file);
ok( PHP::eval_return( "is_uploaded_file('$fake_file')" ),
    "is_uploaded_file true for spoofed fake file" );
PHP::_spoof_rfc1867($fake_file2);
ok( PHP::eval_return( "is_uploaded_file('$fake_file') && is_uploaded_file('$fake_file2')" ),
    "is_uploaded_file true for two spoofed files" );

# 64-71
my @superglobals = qw(_SERVER _GET _POST _FILES _COOKIE _SESSION _REQUEST _ENV);
my $jj = 0;
foreach my $global (@superglobals) {
    $jj++;
    my $data = { foo => 123, bar => 'def', name => $global, jj => $jj };
    PHP::assign_global( $global, $data );
    my $t = PHP::eval_return( '$' . $global );
    ok( $t->{foo} eq '123' && $t->{bar} eq 'def' && $t->{name} eq $global && $t->{jj} == $jj,
	"assignment to superglobal \$$global");
}

# 72-75
my $test_file = "upload_test1";
my $test_file2 = "upload test2";
unlink $test_file, $test_file2;
ok( ! PHP::call('move_uploaded_file', $test_file, $test_file2),
    'move_uploaded_file not successful yet' );
open my $fh, '>', $test_file;
print $fh "upload test file\n";
close $fh;
PHP::_spoof_rfc1867( $test_file );
ok( -f $test_file && PHP::eval_return( "is_uploaded_file('$test_file')" ),
    'test upload file acknowledged by PHP' );
ok( PHP::call('move_uploaded_file', $test_file, $test_file2),
    'move_uploaded_file successful on spoofed upload' );
ok( -f $test_file2 && ! -f $test_file,
    'moved uploaded file has a new name' );
unlink $test_file, $test_file2;

# 76
PHP::set_php_input("0123456789012345678\n" x 568);
my $t76 = PHP::eval_return( 'file_get_contents("php://input")' );
ok( $t76 eq "0123456789012345678\n" x 568,
    'post content avail in php://input' );


PHP.pm
package PHP;

# $Id: PHP.pm,v 1.28 2011/07/26 07:35:13 dk Exp $

use strict;
require DynaLoader;
use vars qw($VERSION $v5 @ISA);
@ISA = qw(DynaLoader);

# remove this or change to 0x00 of your OS croaks here
sub dl_load_flags { 0x01 }

$VERSION = '0.16';
bootstrap PHP $VERSION;

PHP::options( debug => 1) if $ENV{P5PHPDEBUG}; 
$v5 = 1 if PHP::options( 'version') =~ /^(\d+)/ and $1 > 4;

sub END
{
	&PHP::done();
}

sub call	{ PHP::exec( 0, @_) }
sub include	{ PHP::eval( "include('$_[0]');") }
sub require	{ PHP::eval( "require('$_[0]');") }
sub include_once{ PHP::eval( "include_once('$_[0]');") }
sub require_once{ PHP::eval( "require_once('$_[0]');") }
sub __reset     { no warnings 'redefine'; PHP::done(); PHP::_reset(); }

sub array       { PHP::Array-> new(shift) }

our %_seen_zvals;

sub assign_global
{
	my ($varname, $value) = @_;
	local %_seen_zvals;
	if ($varname eq '_REQUEST' || $varname eq '_SERVER' || $varname eq '_ENV') {
		# don't know why this works, but assignment to these superglobals won't
		# take without this step.
		PHP::eval( "\$$varname;" );
	}
	PHP::_assign_global($varname, _to_zval($value));
}

sub _to_zval
{
	require Scalar::Util;
	my $value = shift;

	my $reftype = Scalar::Util::reftype($value);
	$value = undef if ref(\$value) eq 'GLOB';
	return $value unless $reftype;

	return $_seen_zvals{"$value"} if exists $_seen_zvals{"$value"};

	if ( $reftype eq 'SCALAR') {
		$_seen_zvals{"$value"} = undef;
		return $_seen_zvals{"$value"} = _to_zval($$value);
	} elsif ( $reftype eq 'ARRAY') {
		my $zval = PHP::array;
		$_seen_zvals{"$value"} = $zval;
		$zval->[$_] = _to_zval($value->[$_]) for 0 .. $#$value;
		return $zval;
	} elsif ( $reftype eq 'HASH') {
		my $zval = PHP::array;
		$_seen_zvals{"$value"} = $zval;
		$zval->{$_} = _to_zval($value->{$_}) for keys %$value;
		return $zval;
	} else {
		return undef;
	}
}

my $LOADED = 1;

sub AUTOLOAD
{
	die "Module PHP failed to load" unless $LOADED;
	no strict;
	my $method = $AUTOLOAD;
	$method =~ s/^.*://;
	PHP::exec( 0, $method, @_);
}

package PHP::Entity;

sub CREATE
{
	my $class = shift;
	my $self = {};
	bless( $self, $class);
	return $self;
}

sub tie
{
	my ( $self, $tie_to) = @_;
	if ( ref( $tie_to) eq 'HASH') {
		tie %$tie_to, 'PHP::TieHash', $self;
	} elsif ( ref( $tie_to) eq 'ARRAY') {
		tie @$tie_to, 'PHP::TieArray', $self;
	} else {
		die "PHP::Entity::tie: Can't tie to ", ref($tie_to), "\n";
	}
}

package PHP::Object;
use vars qw(@ISA);
@ISA = qw(PHP::Entity);

sub new
{
	my ( $class, $php_class, @params) = @_;
	my $self = $class-> _new( $php_class);
	if ( PHP::exec( 0, 'method_exists', $self, $php_class)) {
		PHP::exec( 1, $php_class, $self, @params)
	} elsif ( $PHP::v5 and PHP::exec( 0, 'method_exists', $self, '__construct')) {
		PHP::exec( 1, '__construct', $self, @params)
	}
	return $self;
}

sub AUTOLOAD
{
	no strict;
	my $method = $AUTOLOAD;
	$method =~ s/^.*://;
	PHP::exec( 1, $method, @_);
}

package PHP::ArrayHandle;
use vars qw(@ISA);
@ISA = qw(PHP::Entity);

package PHP::TieHash;

sub TIEHASH
{
	my ( $class, $self) = @_;
	my $alias = {};
	PHP::Entity::link( $self, $alias);
	return bless( $alias, $class);
}

sub UNTIE
{
	PHP::Entity::unlink( $_[0] );
}

sub DESTROY { goto &UNTIE }

package PHP::TieArray;

sub TIEARRAY
{
	my ( $class, $self) = @_;
	my $alias = {};
	PHP::Entity::link( $self, $alias);
	return bless( $alias, $class);
}

sub UNTIE
{
	PHP::Entity::unlink( $_[0] );
}

sub EXTEND {}
sub STORESIZE {}
sub DESTROY { goto &UNTIE }

package PHP::Array;

my ( %instances);

use overload 
	'%{}' => sub { $instances{PHP::stringify($_[0])}->[0] },
	'@{}' => sub { $instances{PHP::stringify($_[0])}->[1] },
	'""'  => sub { PHP::stringify($_[0]) };

sub new
{
	my ( $class, $handle) = @_;
	$handle = PHP::ArrayHandle-> new unless $handle;
	my ( $self, $hash_instance, $array_instance) = 
		( {}, {}, []);
	my $id = PHP::stringify( $self);
	$instances{$id} = [ $hash_instance, $array_instance, $handle ];
	tie %$hash_instance, 'PHP::TieHash', $handle;
	tie @$array_instance, 'PHP::TieArray', $handle;
	PHP::Entity::link( $handle, $self);
	bless ( $self, $class);
	return $self;
}

sub handle { $instances{"$_[0]"}->[2] }

sub tie
{
	my ( $self, $tie_to) = @_;
	if ( ref( $tie_to) eq 'HASH') {
		tie %$tie_to, 'PHP::TieHash', $self-> handle;
	} elsif ( ref( $tie_to) eq 'ARRAY') {
		tie @$tie_to, 'PHP::TieArray', $self-> handle;
	} else {
		die "PHP::Array::tie: Can't tie to ", ref($tie_to), "\n";
	}
}

sub DESTROY
{
	my $self = $_[0];
	PHP::Entity::unlink( $self);
	delete $instances{ PHP::stringify( $self)};
}

1;

__DATA__

=pod

=head1 NAME

PHP - embedded PHP interpreter

=head1 DESCRIPTION

The module makes it possible to execute PHP code, call PHP functions and methods,
manipulate PHP arrays, and create PHP objects.

=head1 SYNOPSIS

	use PHP;

General use

	# evaluate arbitrary PHP code; exception is thrown
	# and can be caught via standard eval{}/$@ block 
	PHP::eval(<<'EVAL');
	function print_val($arr,$val) {
		echo $arr[$val];
	}
	
	class TestClass {
		function TestClass ( $param ) {}
		function method($val) { return $val + 1; }
	};
	EVAL

	# catch output of PHP code
	PHP::options( stdout => sub {
		print "PHP says: $_[0]\n";
	});
	PHP::eval('echo 42;');

Arrays, high level

	# create a php array
	my $array = PHP::array;

	# access pseudo-hash content
	$array-> [1] = 42;
	$array-> {string} = 43;
	
	# pass arrays to function
	# Note - function name is not known by perl in advance, and
	# is called via AUTOLOAD
	PHP::print_val($array, 1);
	PHP::print_val($array, 'string');

Arrays, low level

	# create a php array handle
	my $array = PHP::ArrayHandle-> new();
	# tie it either to an array or a hash
	my ( @array, %hash);
	$array-> tie(\%hash);
	$array-> tie(\@array);

	# access array content
	$array[1] = 42;
	$hash{2} = 43;

Objects and properties

	my $TestClass = PHP::Object-> new('TestClass');
	print $TestClass-> method(42), "\n";
	
	$TestClass-> tie(\%hash);
	# set a property
	$hash{new_prop} = 'string';

=head1 API

=over

=item eval $CODE

Feeds embedded PHP interpreter with $CODE, throws an exception on
failure. This method does not have a return value. See also C<eval_return>.

=item eval_return $CODE

Same as C<eval> but returns the calculated value. This method can be used 
for any expression where it would make sense if you put a C<"return "> in
front of it. Otherwise you should use C<eval>.

    $x = PHP::eval_return("13*86;");                           # ok
    $x = PHP::eval_return('$var + func_that_returns_val();');  # ok
    $x = PHP::eval_return('$var < 0 ? $var : array(7,8,9);');  # ok
    $x = PHP::eval_return('function foo() { return 75;}');     # not ok
    $x = PHP::eval_return('if ($var<0) { $bar=$foo; ');        # not ok
    $x = PHP::eval_return('echo "This is a message";');        # not ok



The PHP interpreter does an implicit prepend
of "return " text to C<$CODE>, so beware.

=item call FUNCTION ...

Calls PHP function with list of parameters. 
Returns exactly one value.

=item include, include_once, require, require_once

Shortcuts to the identical PHP constructs.

=item assign_global NAME, VALUE

Assigns the given VALUE to the global PHP variable C<$NAME>,
converting Perl data types to PHP data types as necessary.
VALUE may be a list reference, hash reference, scalar reference,
or regular scalar.

=item array [ $REFERENCE ]

Returns a handle to a newly created C<PHP::Array> object, which 
can be accessed both as array and hash reference:

	$_ = PHP::array;
	$_->[42] = 'hello';
	$_->{world} = '!';

If $REFERENCE is a C<PHP::ArrayHandle> instance, then the newly created object
is a pheudo-hash alias to the PHP array behind the $REFERENCE. If no 
$REFERENCE is given, a new PHP array is created.

=item PHP::Object->new($class_name, @parameters)

Instantiates a PHP object of PHP class $class_name and returns a handle to it.
The methods of the class can be called directly via the handle:

	my $obj = PHP::Object-> new( 'MyClass', @params_to_constructor);
	$object-> method( @some_params);


The relevant class constructor is called, if available, according to PHP
specification, that is different between v4 and v5. The v4 constructor has
identical name with the class name; the v5 constructor can also be named
C<__construct>.

=item PHP::Entity->tie($array_handle, $tie_to)

Ties existing handle to a PHP entity to either a perl hash or a perl array.
The tied hash or array can be used to access PHP pseudo_hash values indexed
either by string or integer value. 

The PHP entity can be either an array, represented by C<PHP::ArrayHandle>, or
an object, represented by C<PHP::Object>. In the latter case, the object 
properties are represented as hash/array values.

=item PHP::Entity->link($original, $link)

Records a reference to an arbitrary perl scalar $link as an
alias to $original C<PHP::Entity> object. This is used internally
by C<PHP::TieHash> and C<PHP::TieArray>, but might be also used
for other purposes.

=item PHP::Entity::unlink($link)

Removes association between a C<PHP::Entity> object and $link.

=item PHP::Array->tie($self, $tie_to)

Same as L<< PHP::Entity->tie >>, but operates on C<PHP::Array> objects.

=item PHP::Array->handle

Returns PHP array handle, a C<PHP::ArrayHandle> object.

=item PHP::options

Contains set of internal options. If called without parameters,
returns the names of the options. If called with a single parameter,
return the associated value. If called with two parameters, replaces
the associated value.

=over

=item debug $integer

If set, loads of debugging information are dumped to stderr

Default: 0

=item stdout/stderr $callback

C<stdout> and C<stderr> options define callbacks that are called
when PHP decides to print something or complain, respectively.

Default: undef

=item header $callback

Callback when PHP sets a response header with the PHP C<header()>
function. The callback will receive two arguments, corresponding to 
the first two arguments of the PHP C<header()> function. 

Default: undef

=item version

Read-only option; returns the version of PHP library compiled with .

=back

=item PHP::set_php_input($string)

Sets content for PHP applications that read from the C<<php://input>
stream.

=item PHP::_spoof_rfc1867($filename)

Manipulates an internal hash in PHP so that PHP's C<is_uploaded_file>
function will return true for the given filename. This can be helpful 
when you have already manipulated PHP's global C<$_FILES> variable
for an application that uploads files. But it is not always necessary.

=back

=head1 DEBUGGING

Environment variable C<P5PHPDEBUG>, if set to 1, turns the debug mode on. The
same effect can be achieved programmatically by calling

	PHP::options( debug => 1);

=head1 INSTALLATION

The module uses php-embed SAPI extension to inter-operate with PHP interpreter.
That means php must be configured with '--enable-embed' parameters prior to
using the module. Also, no '--with-apxs' must be present in to configuration
agruments either, otherwise the PHP library will be linked with Apache functions,
and will be unusable from the command line.

The C<sub dl_load_flags { 0x01 }> code in F<PHP.pm> is required for PHP
to load correctly its extensions. If your platform does RTLD_GLOBAL by
default and croaks upon this line, it is safe to remove the line.

=head1 WHY?

While I do agree that in general it is absolutely pointless to use PHP
functionality from within Perl, scenarios where one must connect an existing
PHP codebase to something else, are not something unusual. Also, this module
might be handy for people who know PHP but are afraid of switching to Perl, or want to
reuse their old PHP code.

Currently, not all of PHP functionality is implemented, but OTOH I don't really
expect this module to grow that big, because I believe it is easier to call
C<PHP::eval> rather than implement all the subtleties of Zend API. There are no
callbacks to Perl from PHP code, and I don't think these are needed, because
one thing is to be lazy and not to rewrite PHP code, and another is to make new
code in PHP that uses Perl when PHP is not enough. As I see it, the latter
would kill all incentive to switch to Perl, so I'd rather leave callbacks
unimplemented.

=head1 SEE ALSO

Using Perl code from PHP: L<http://www.zend.com/php5/articles/php5-perl.php>

=head1 COPYRIGHT

This library is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

=head1 AUTHOR

Dmitry Karasik <dmitry@karasik.eu.org>

=cut
